### ISSUE TITLE
Process Expansion State

### ISSUE THREAD

**Issue by nickeb96** [fish-shell/fish-shell#9811]
Can the `%self` syntax be removed from fish yet?  From what I understand, there used to be several `%` expansions and all of them were removed, except for `%self` and I'm not sure why.

Is there any difference between `%self` and `$fish_pid`?  Should one be preferred over the other?  The percent version feels out of place.

**Comment by faho** [comment]
>Can the %self syntax be removed from fish yet?

There is no need to rush this.

It's quite out of the way, it doesn't hurt to keep it.

>From what I understand, there used to be several % expansions and all of them were removed

The expansion was `%something`, and what happened was fish would look for a process with "something" in its commandline and return the PID. This wasn't very useful, mostly because there was no way to catch failure (if it didn't match it would print an error), and because fish could only ever look at its own children.

So it *looked* like you could do `kill %firefox`, but that only worked if you had started firefox *in that fish*, with `firefox &`.

Also it integrated quite terribly with the rest, because you couldn't even do `/proc/%firefox`, because the `%` needed to be at the beginning of a word.

So %-expansion was almost entirely unused, and made `%` a character to watch out for.

>except for %self and I'm not sure why

%self *was* actually used, mostly because $fish_pid didn't exist. `%self` also can't *fail*, so there's less of a problem with it specifically. And so we kept it, to avoid a flag day transition in quite a bunch of plugins and such.

>Is there any difference between %self and $fish_pid? Should one be preferred over the other?

See above. The latter is simpler and integrates better. It is supported since fish 3.0, released in February 2019.

**Comment by faho** [comment]
Okay, question was answered, closing.

**Comment by nickeb96** [comment]
Could we at least make removing `%self` a long term goal?  Fish 4 or something?

The `%self` syntax feels like it's in the same relic category as the `cmd args ^/dev/null` syntax for stderr redirection.

**Comment by faho** [comment]
>Could we at least make removing %self a long term goal? Fish 4 or something?

It *is* a long term "goal". Even that is saying a bit much, removing it is extremely easy, it's just that there's no need to rush.

Nothing forces you to use it, it's not particularly prominent or in the way. It's just a teensy bit of backwards compatibility.

**Comment by nickeb96** [comment]
> For comparison, that actually was in the way - it meant that ^ was a special character pretty much always, and it made things like git reset ^ annoying.

Keeping `%self` means that `%` remains a special character as well.  It may not be quite as intrusive as `^`, but you still need to escape it anytime you need a literal percent and you can't be 100% sure of what follows it.

Backwards compatibility is important, but I still think that it should be removed or put behind a feature flag next time there is a breaking update.  I still don't see how this is any different from changes made to `^`, `?`, or `&` (without whitespace around it).

> And the 2> replacement syntax existed for ages before that (while $fish_pid was added in 3.0).

Fish 3.0 was released in 2018, so roughly 5 years with `$fish_pid` being preferred.  I don't think that's an unreasonably short amount time to introduce a deprecation plan.

**Comment by faho** [comment]
>you still need to escape it anytime you need a literal percent and you can't be 100% sure of what follows it.

If you can't be 100% sure what follows it, it can't be a literal `%self`.

If it was a literal `%self`, you would know that it was that.

**Comment by nickeb96** [comment]
Sure, let me rephrase it:

> You still need to escape it sometimes, because it is still in fact a special character.

Why keep this relic syntax?  It very clearly goes against [the law of orthogonality](https://fishshell.com/docs/current/design.html#the-law-of-orthogonality) in the design docs:

> The shell language should have a small set of orthogonal features. Any situation where two features are related but not identical, one of them should be removed, and the other should be made powerful and general enough to handle all common use cases of either feature.

**Comment by faho** [comment]
I deleted my comment because I thought that way I could let it lie, but it appears my timing was off, sorry about that.

I have explained our position, I believe that should be enough. I do not believe there is more to discuss here.

**Comment by faho** [comment]
Just to be clear: We plan on eventually removing `%self`. We have always planned on eventually removing it, and that's why I told you it was a long term goal, and why I said "there is no need to rush this" as opposed to "no, we don't want to remove it".

The removal won't have a feature flag, and it won't really be in a major release.

We've had our experience with the `^` removal, which we pressed quite aggressively, and it wasn't great - a bunch of semi-maintained plugins and such were broken by it, even as we called it out, and added the feature flag, and called it out again.

So we'll just take this one nice and slow, especially because there is no compelling reason to do it any faster other than vague waving at PHILOSOPHY.

There's a reason the Zen Of Python says "practicality beats purity".

I hope that should settle it?

**Comment by nickeb96** [comment]
I would appreciate it if you would reopen my issue so other members of the community and developer group, such as @ridiculousfish, could contribute to the discussion.  Some of the problems I've brought up have not been addressed or discussed.  I'll reiterate them below:


1.  Keeping both `%self` and `$fish_pid` goes against [the law of orthogonality](https://fishshell.com/docs/current/design.html#the-law-of-orthogonality) from the design docs.  Temporarily having both for the sake of backwards compatibility is fine, but it's been 5 years and proposing a deprecation plan for `%self` is not unreasonable.
    
    Also, the backwards compatibility argument is less relevant now since most scripts from pre-2018 would not run on current versions of fish anyway, regardless of `%self`'s status.


2.  Feature gates have been added for changing `^`, `?`, and `&` (in a token) for similar reasons.  Why is there a problem with adding a similar opt-in feature for `%self` and how would that be any different from the other three?

**Comment by faho** [comment]
Ah, that's terrible timing. Please see [my comment](https://github.com/fish-shell/fish-shell/issues/9811#issuecomment-1574286195) above, it appears you have entirely misunderstood my communication

**Comment by nickeb96** [comment]
Yes, your reply loaded in just as I was submitting mine.  I think the difference with `^` is that `^` removal was on by default iirc.  I think that if `%self` removal became an opt-in feature initially, no one would complain.

**Comment by faho** [comment]
There is no need to make it an opt-in feature. We'll just remove it, *eventually*.

That is our position, that is what I've been saying, and I think that's enough, now.

I would like if we could keep it at that.

**Comment by krobelus** [comment]
feature flags exist to give enterprising users early access to upcoming changes.
Here this would help whoever randomly happens to use `%self`, which is quite unlikely, probably not worth the noise of a feature flag :)
